src/data/matrix/basic.lean

```diff
@@ -95,33 +106,98 @@ end diagonal
+def dot_product [has_mul α] [add_comm_monoid α] (v w : m → α) : α :=
+∑ i, v i * w i
+
+lemma dot_product_assoc [semiring α] (u : m → α) (v : m → n → α) (w : n → α) :
+  dot_product (λ j, dot_product u (λ i, v i j)) w = dot_product u (λ i, dot_product (v i) w) :=
+by simpa [dot_product, finset.mul_sum, finset.sum_mul, mul_assoc] using finset.sum_comm
+
+lemma dot_product_comm [comm_semiring α] (v w : m → α) :
+  dot_product v w = dot_product w v :=
+by simp_rw [dot_product, mul_comm]
+
+@[simp] lemma dot_product_punit [add_comm_monoid α] [has_mul α] (v w : punit → α) :
+  dot_product v w = v ⟨⟩ * w ⟨⟩ :=
+by simp [dot_product]
+
+@[simp] lemma dot_product_zero [semiring α] (v : m → α) : dot_product v 0 = 0 :=
+by simp [dot_product]
+
+@[simp] lemma dot_product_zero' [semiring α] (v : m → α) : dot_product v (λ _, 0) = 0 :=
+dot_product_zero v
+
+@[simp] lemma zero_dot_product [semiring α] (v : m → α) : dot_product 0 v = 0 :=
+by simp [dot_product]
+
+@[simp] lemma zero_dot_product' [semiring α] (v : m → α) : dot_product (λ _, (0 : α)) v = 0 :=
+zero_dot_product v
+
+@[simp] lemma add_dot_product [semiring α] (u v w : m → α) :
+  dot_product (u + v) w = dot_product u w + dot_product v w :=
+by simp [dot_product, add_mul, finset.sum_add_distrib]
+
+@[simp] lemma dot_product_add [semiring α] (u v w : m → α) :
+  dot_product u (v + w) = dot_product u v + dot_product u w :=
+by simp [dot_product, mul_add, finset.sum_add_distrib]
+
+@[simp] lemma diagonal_dot_product [decidable_eq m] [semiring α] (v w : m → α) (i : m) :
+  dot_product (diagonal v i) w = v i * w i :=
+have ∀ j ≠ i, diagonal v i j * w j = 0 := λ j hij, by simp [diagonal_val_ne' hij],
+by convert finset.sum_eq_single i (λ j _, this j) _; simp
+
+@[simp] lemma dot_product_diagonal [decidable_eq m] [semiring α] (v w : m → α) (i : m) :
+  dot_product v (diagonal w i) = v i * w i :=
+have ∀ j ≠ i, v j * diagonal w i j = 0 := λ j hij, by simp [diagonal_val_ne' hij],
+by convert finset.sum_eq_single i (λ j _, this j) _; simp
+
+@[simp] lemma dot_product_diagonal' [decidable_eq m] [semiring α] (v w : m → α) (i : m) :
+  dot_product v (λ j, diagonal w j i) = v i * w i :=
+have ∀ j ≠ i, v j * diagonal w j i = 0 := λ j hij, by simp [diagonal_val_ne hij],
+by convert finset.sum_eq_single i (λ j _, this j) _; simp
+
+@[simp] lemma neg_dot_product [ring α] (v w : m → α) : dot_product (-v) w = - dot_product v w :=
+by simp [dot_product]
+
+@[simp] lemma dot_product_neg [ring α] (v w : m → α) : dot_product v (-w) = - dot_product v w :=
+by simp [dot_product]
+
+@[simp] lemma smul_dot_product [semiring α] (x : α) (v w : m → α) :
+  dot_product (x • v) w = x * dot_product v w :=
+by simp [dot_product, finset.mul_sum, mul_assoc]
+
+@[simp] lemma dot_product_smul [comm_semiring α] (x : α) (v w : m → α) :
+  dot_product v (x • w) = x * dot_product v w :=
+by simp [dot_product, finset.mul_sum, mul_assoc, mul_comm, mul_left_comm]
+
+end dot_product
```

Comment:
I think this is good. The proofs below become nice and short.
One tiny comment: can you link this to a product of a `row` and `col` vector?

Reply:
I have added [`row_mul_col_val` here](https://github.com/leanprover-community/mathlib/pull/2656/commits/bcdcfe1e93c82858c99e5e62a625d3b919895f70#diff-d3aacdce67cc97734f3da378a0c338dfR284).

Reply:
Thanks!